File: dotnet/src/Microsoft.AutoGen/Core/PythonEquiv/TypeSubscriptionAttribute.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// TypeSubscriptionAttribute.cs

using Microsoft.AutoGen.Contracts.Python;

namespace Microsoft.AutoGen.Core.Python;

[AttributeUsage(AttributeTargets.All)]
public class TypeSubscriptionAttribute(string topic) : Attribute, IUnboundSubscriptionDefinition
{
    public string Topic { get; } = topic;

    public ISubscriptionDefinition Bind(AgentType agentType)
    {
        return new TypeSubscription(Topic, agentType);
    }
}

File: typescript/core/TypeSubscriptionAttribute.ts
import { ISubscriptionDefinition } from "../contracts/ISubscriptionDefinition";
import { IUnboundSubscriptionDefinition } from "../contracts/IUnboundSubscriptionDefinition";
import { AgentType } from "../contracts/AgentType";
import { TypeSubscription } from "./TypeSubscription";

export class TypeSubscriptionAttribute implements IUnboundSubscriptionDefinition {
  constructor(private readonly topic: string) {}

  bind(agentType: AgentType): ISubscriptionDefinition {
    return new TypeSubscription(this.topic, agentType);
  }
}
